src/basic_graph.py

```diff
@@ -0,0 +1,86 @@
+from graph import Graph
+from model import NodeAttrs, EdgeAttrs, NodeHandle, Edge, Node
+import itertools as it
+
+def basic_square(for_lhs=False, hanging=True):
+    """
+    for_lhs is mostly necessary when basic_* functions are invoked from
+    __create_lhs - then handles are enumerated from 0, which is very
+    handy during processing of monomorphisms.
+    TODO: delete `hanging' argument after P7 is introduced and change it
+    to false
+    """
+    graph = Graph()
+
+    if for_lhs:
+        node_1 = Node(NodeAttrs(label='v', x=0, y=0, flag=False), handle=0)
+        node_2 = Node(NodeAttrs(label='v', x=1, y=0, flag=False), handle=1)
+        node_3 = Node(NodeAttrs(label='v', x=1, y=1, flag=False), handle=2)
+        node_4 = Node(NodeAttrs(label='v', x=0, y=1, flag=False), handle=3)
+        node_q = Node(NodeAttrs(label='q', x=0.5, y=0.5, flag=hanging), handle=4)
+    else:
+        node_1 = Node(NodeAttrs(label='v', x=0, y=0, flag=False))
+        node_2 = Node(NodeAttrs(label='v', x=1, y=0, flag=False))
+        node_3 = Node(NodeAttrs(label='v', x=1, y=1, flag=False))
+        node_4 = Node(NodeAttrs(label='v', x=0, y=1, flag=False))
+        node_q = Node(NodeAttrs(label='q', x=0.5, y=0.5, flag=hanging))
+
+    nodes = [node_1, node_2, node_3, node_4]
+
+    graph.add_node_collection(nodes)
+    graph.add_node(node_q)
+
+    # add `edges' of hyperedge
+    for node in nodes:
+        attr = EdgeAttrs(kind='q', flag=hanging)
+        edge = Edge(node.handle, node_q.handle, attr)
+        graph.add_edge(edge)
```

Comment:
Please use `add_q_hyperedge` here. This code does not ensure single edge handle for all edges that belong to this hyperedge.

Reply:
valid point